### ISSUE TITLE
C++: char arrays potentially read/write too far

### ISSUE THREAD

**Issue by MatthiasKauer** [real-logic/simple-binary-encoding#517]
Hi,
I have observed the following behavior.

Using a char array field like the one from the baseline example (length increased to 20 so it is above initial capacity of std::string on my platform):

    <type name="VehicleCode" primitiveType="char" length="20" characterEncoding="ASCII"/>

```
//main.cpp
#include <iostream>
#include <string>

#include "baseline/Car.h"
#include "baseline/MessageHeader.h"

int main() {
  using namespace baseline;

  std::cout << "hi" << std::endl;

  char buffer[2048];
  MessageHeader hdr;
  Car car;

  car.wrapForEncode(buffer, hdr.encodedLength(), sizeof(buffer));
  car.putVehicleCode("ab");
  std::cout << car.getVehicleCodeAsString() << std::endl;
  std::cout << car.vehicleCode() << std::endl;

  std::string s = "SOMEVALUES";

  std::cout << "bye" << std::endl;
}
```

Output:
```
o A2F6D matt@hp840:~/dvl/sbe_bug $ g++ -g -O2 -I ~/dvl/sbe_bug -std=c++14 main.cpp && ./a.out 
hi
abSOMEVALUESbye
ab
bye
```

Valgrind also finds errors. The problem seems to be in the following generated code:
```

    Car &putVehicleCode(const char *src)
    {
        std::memcpy(m_buffer + m_offset + 32, src, sizeof(char) * 20);
        return *this;
    }

    std::string getVehicleCodeAsString() const
    {
        std::string result(m_buffer + m_offset + 32, 20);
        return result;
    }

    Car &putVehicleCode(const std::string& str)
    {
        std::memcpy(m_buffer + m_offset + 32, str.c_str(), 20);
        return *this;
    }
```
The length of the string could be checked. I am not sure what the best approach is for the `char *` setter.

Are you interested in a pull request for this?

Best regards,
Matthias

**Comment by tmontgomery** [comment]
Did you turn off bounds checking?

**Comment by tmontgomery** [comment]
Also, those setters assume the app is checking termination. It is only protecting the underlying buffer.

Also, the field is a set length in size. It's 20 bytes. If the app passes less than that, it is the apps fault that it didn't adhere to the format of the field.

**Comment by MatthiasKauer** [comment]
I believe bounds checking is on by default and I left it that way. My compilation options are quite short and included above.

> Also, the field is a set length in size. It's 20 bytes. If the app passes less than that, it is the apps fault that it didn't adhere to the format of the field.

One can see it that way and then there is really nothing to do.
My suggestion would have been to check the length of the string and maybe adjust the other setter to include the length of the char buffer from the call site, i.e, 

```

    Car &putVehicleCode(const char *src, size_t len)
    {
        if (len < 20) {
            some_error();
        }
        std::memcpy(m_buffer + m_offset + 32, src, sizeof(char) * 20);
        return *this;
    }
```
Of course one may have to think about such a change a bit more, maybe make the checks optional.

**Comment by tmontgomery** [comment]
Another option (for C++11) is to pass a reference to a `std::array<char, 20>` as a param and thus always force the argument to be of the correct size. However, it might prevent some other usages.... or force users to always use `std::array` which is less than desirable. A `char *` arg is useful.

**Comment by MatthiasKauer** [comment]
I completely agree that `array<char, 20>` is too restrictive. span (from GSL) would be an option but also another dependency. Offering both checked `Car &putVehicleCode(const char *src, size_t len)` and original unchecked `Car &putVehicleCode(const char *src)` might work.

**Comment by tmontgomery** [comment]
Keeping unchecked for speed is a necessity, I am afraid. However, adding a checked version is good. If you want to do a PR, go ahead. Or I will eventually get to it if not.

**Comment by MatthiasKauer** [comment]
I looked at this again today and there has been quite some progress indeed.

The memcpy addition to the getters from https://github.com/real-logic/simple-binary-encoding/issues/506 could change a few things, as you said. I don't see how https://github.com/real-logic/simple-binary-encoding/issues/530 is related but it kind of got fixed in the wake of #506 so maybe that's why you mentioned it as well.

So in summary, we have to re-confirm this issue.
The main objection against https://github.com/real-logic/simple-binary-encoding/pull/522 is that it adds overhead to existing methods and you would prefer an additional safe method that can be opted in, right?

**Comment by tmontgomery** [comment]
@MatthiasKauer correct. There is no need to change existing methods. But adding a method that can be used to check length is fine as I mentioned.
